### ISSUE TITLE
[<Ray component: Core|RLlib|etc...>] ray.remote ray.put OOM

### ISSUE THREAD

**Issue by flytomato2020** [ray-project/ray#48989]
### What happened + What you expected to happen

I have several task groups, each of which ([[0,1,2,3], "post_process1"]) consists of n gpu tasks and one cpu task. The cpu task requires the return value of the gpu task. Let's say I'm running on a two-card server. When I run tasks 0,1, I store the result using ray.put. At this point ray starts tasks 2,3. But the video memory is not freed. This leads to OOM. How should I handle this

### Versions / Dependencies

ray   2.39.0

### Reproduction script

import os
import socket
import sys

os.environ["RAY_DEDUP_LOGS"] = '0'
import ray
import time
import torch
import uuid
import traceback
import threading
import numpy as np
from queue import Queue
from flask import Flask

# 初始化 ray
ray.init()

# 创建 Flask 应用
app = Flask(__name__)

# 定义任务组队列
TaskQueue = Queue()

# 定义GPU密集型子任务（自动由Ray调度到空闲的GPU上）
@ray.remote(num_gpus=1)
def gpu_task(i):
    # visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES", "0")
    try:
        print(i)
        # 设置指定 GPU
        gpu_gb = 10
        cpu_mb = 1000
        time_s = 20
        # 1GB = 1024 ** 3 bytes
        bytes_in_gb = 1024 ** 3
        target_bytes = gpu_gb * bytes_in_gb

        # 每个 float32 变量占用 4 bytes
        num_elements = target_bytes // 4

        # 在指定的 GPU 上分配大张量
        tensor = torch.empty(num_elements, dtype=torch.float32, device='cuda')
        # print(f"Allocated tensor of shape {tensor.shape} on GPU:{gpu_id}, approximately {gpu_gb} GB of memory.")

        number_of_floats = int(cpu_mb * 1024 * 1024 // 8)
        memory_hog = [0.0] * number_of_floats
        # print(f"Allocated {cpu_mb} MB of memory.")
        time.sleep(time_s)
        mask = tensor.cpu().numpy()
        mask_ref = ray.put(mask)  # 将数据放入共享内存

        return mask_ref
    except Exception as e:
        error_message = traceback.format_exc()
        print(f"Error in task {str(error_message)}")
        raise e

# 定义后处理任务（使用 CPU）
@ray.remote
def post_process(*mask_refs):
    try:
        # 从共享内存中获取数据并进行处理
        for mask_ref in mask_refs:
            mask = ray.get(mask_ref)

        print('post_process end !!!!!!!!!!!!!!!!!!!!!!!!!!!! ')
        return f"Post-processing result"
    except Exception as e:
        error_message = traceback.format_exc()
        print(f"Error in post_process: {str(error_message)}")
        raise e

class TaskScheduler:
    def __init__(self):
        self.gpu_futures = []
        self.post_tasks = []
        self.in_progress_task_groups = {}

    def task_scheduler(self):
        """任务调度和监控函数"""
        while True:
            while not TaskQueue.empty() or self.gpu_futures:
                # 启动新的任务组（如果有空闲的 GPU）
                if not TaskQueue.empty():
                    next_task_group = TaskQueue.get()
                    subtasks, post_task_name = next_task_group

                    print(f"启动任务组 {subtasks}...")
                    futures = [gpu_task.remote(i) for i in subtasks]
                    self.gpu_futures.extend(futures)
                    self.in_progress_task_groups[post_task_name] = futures

                # 检查是否有完成的 GPU 任务，并执行后处理任务
                if self.gpu_futures:
                    done_ids, self.gpu_futures = ray.wait(self.gpu_futures, num_returns=1, timeout=0.1)
                    if done_ids:
                        for done_id in done_ids:
                            try:
                                result = ray.get(done_id)  # 获取任务结果，如果任务出错则会在这里抛出异常
                                print(f"GPU Task Completed: {result}")
                            except Exception as e:
                                print(f"Error detected, stopping current task group: {str(e)}")
                                # 如果任务出错，停止当前任务组中的其他子任务
                                for post_task_name, task_list in self.in_progress_task_groups.items():
                                    if done_id in task_list:
                                        # 将当前任务组的所有其他子任务从 gpu_futures 中移除
                                        for task in task_list:
                                            if task in self.gpu_futures:
                                                self.gpu_futures.remove(task)
                                        # 从进行中的任务组中移除
                                        self.in_progress_task_groups.pop(post_task_name)
                                        break

                        # 检查是否某任务组的所有子任务都已完成，启动后处理任务
                        completed_task_groups = []
                        for post_task_name, task_list in self.in_progress_task_groups.items():
                            if all(task in done_ids or task not in self.gpu_futures for task in task_list):
                                # 启动对应的后处理任务
                                print(f"启动后处理任务 {post_task_name}...")
                                post_task = post_process.remote(*task_list)
                                self.post_tasks.append(post_task)
                                completed_task_groups.append(post_task_name)

                        # 从进行中的任务组中移除已完成的任务组
                        for post_task_name in completed_task_groups:
                            print(post_task_name)
                            self.in_progress_task_groups.pop(post_task_name)

    def start_scheduler(self):
        """启动任务调度器并运行在后台"""
        scheduler_thread = threading.Thread(target=self.task_scheduler)
        scheduler_thread.start()

scheduler = TaskScheduler()
main_thread = threading.Thread(target=scheduler.task_scheduler, daemon=True)
main_thread.start()

if __name__ == '__main__':
    TaskQueue.put([[0,1,2,3], "post_process1"])
    TaskQueue.put([[4,5,6,7], "post_process2"])
    while True:
        time.sleep(10)

### Issue Severity

High: It blocks me from completing my task.

**Comment by jjyao** [comment]
Do you see CPU memory OOM or GPU memory OOM? `ray.put()` saves data to plasma store which is CPU memory.
